indexing-service/src/main/java/org/apache/druid/indexing/overlord/TaskLockbox.java

```diff
@@ -494,10 +494,11 @@ public List<SegmentAllocateResult> allocateSegments(
         allocateSegmentIds(dataSource, interval, skipSegmentLineageCheck, holderList.getPending());
         holderList.getPending().forEach(holder -> acquireTaskLock(holder, false));
       }
-
       holderList.getPending().forEach(holder -> addTaskAndPersistLocks(holder, isTimeChunkLock));
     }
     finally {
+      holderList.failHoldersForCompletedTasks(this);
+      holderList.clearStaleLocks(this);
```

Comment:
failHoldersForCompletedTasks may help prevent retries of pending requests after the task itself has stopped even when there is no exception.
clearStaleLocks is not needed but I don't think it can cause any harm.

Reply:
@abhishekagarwal87, thanks, I have made changes to clear the locks only in the case of exceptions.
I'm also omitting the changes to stop retries for completed tasks from this PR